### PR TITLE
fix(cargo): Bump complex ranges

### LINKED ISSUE #4765: Cargo bump fails with complex ranges
Repo: https://github.com/mozilla/fxa

Error:
```
    "repository": "mozilla/fxa",
    "err": {
        "message": "Cannot read property 'split' of null",
        "stack": "TypeError: Cannot read property 'split' of null\n    at npm2cargo (/home/ubuntu/renovateapp/node_modules/renovate/dist/versioning/cargo/index.js:25:10)\n    at Object.getNewValue (/home/ubuntu/renovateapp/node_modules/renovate/dist/versioning/cargo/index.js:58:20)\n    at lookupUpdates (/home/ubuntu/renovateapp/node_modules/renovate/dist/workers/repository/process/lookup/index.js:126:43)\n    at process._tickCallback (internal/process/next_tick.js:68:7)"
    },
    "currentValue": ">=0.1.21,<0.2",
    "rangeStrategy": "bump",
    "fromVersion": "0.1.21",
    "toVersion": "0.1.24",
```

### PR DESCRIPTION
**Pull request by zharinov**
Closes #4765

### CODE REVIEW

**Review by viceice — COMMENTED** [review]
Doesn't this affect npm versioning too?

**Comment by rarkins** [comment]
Does it really belong in npm or do we need to refactor cargo a little before fixing this?

**Comment by zharinov** [comment]
Test can be translated to `npm`, then little modifications in `cargo` should finally resolve original issue

File: lib/versioning/npm/range.ts [hunk 1]
```diff
@@ -113,6 +113,39 @@ export function getNewValue({
...
+        }
+        const replaced = getNewValue({
+          currentValue: subRange,
+          rangeStrategy: 'replace',
+          fromVersion,
+          toVersion: incremented,
+        });
+        return satisfies(toVersion, replaced) ? replaced : null;
+      });
+      return versions.filter((x) => x !== null && x !== '').join(', ');
```
Comment (rarkins) [inline lib/versioning/npm/range.ts:148]: I think commas aren't valid in semver ranges, so this is incorrect to have in npm versioning

**Review by rarkins — CHANGES_REQUESTED** [review]

**Comment by rarkins** [comment]
So will this fix some npm complex ranges too?

**Comment by zharinov** [comment]
> So will this fix some npm complex ranges too?

Now should bump to the new result instead of `null`, but can you please provide more examples with what would you like to see?

**Comment by zharinov** [comment]
Example repo: https://github.com/zharinov/fxa/pulls

**Comment by zharinov** [comment]
It's current behavior is subject to discussion I think

File: lib/versioning/npm/index.spec.ts [hunk 2]
```diff
@@ -242,4 +242,78 @@ describe('semver.getNewValue()', () => {
...
+      semver.getNewValue({
+        currentValue: '>= 1.2.3 <= 1.2.0',
+        rangeStrategy: 'bump',
+        fromVersion: '1.2.3',
+        toVersion: '1.2.4',
+      })
+    ).toEqual('>= 1.2.4 <= 1.3.0');
+    expect(
+      semver.getNewValue({
+        currentValue: '>= 1.2.3 <= 1.0',
```
Comment (rarkins) [inline lib/versioning/npm/index.spec.ts:296]: Isn't this test case invalid?

File: lib/versioning/npm/index.spec.ts [hunk 3]
```diff
@@ -242,4 +242,78 @@ describe('semver.getNewValue()', () => {
...
+        rangeStrategy: 'bump',
+        fromVersion: '1.2.3',
+        toVersion: '1.2.4',
+      })
+    ).toEqual('>= 1.2.4 <= 1.2');
+    expect(
+      semver.getNewValue({
+        currentValue: '>= 1.2.3 <= 1.0.0',
```
Comment (rarkins) [inline lib/versioning/npm/index.spec.ts:304]: Also invalid?

**Review by rarkins — CHANGES_REQUESTED** [review]
Review invalid range tests

**Comment by zharinov** [comment]
With all these invalid test inputs I wanted to show it produces correct ranges from it anyway. But let's keep this simpler and just collect user feedback on edge-cases.
